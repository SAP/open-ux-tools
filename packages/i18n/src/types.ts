import type { Range } from './parser/utils';

/**
 * Value node
 */
export interface ValueNode<T> {
    value: T;
    range: Range;
}

/**
 * Text node
 */
export type TextNode = ValueNode<string>;

/**
 * i18n annotation node
 */
export interface I18nAnnotationNode {
    textType: ValueNode<SapTextType>;
    maxLength?: ValueNode<number>;
    /**
     * Note for translator
     */
    note?: TextNode;
}

/**
 * Annotations for the translation entry, which can contain additional metadata about the entry.
 */
export interface I18nAnnotation {
    textType: SapTextType;
    maxLength?: number;
    /**
     * Note for translator
     */
    note?: string;
}

/**
 * new i18n entry
 */
export interface NewI18nEntry {
    key: string;
    value: string;
    /**
     * Annotations for the translation entry, which can contain additional metadata about the entry
     */
    annotation?: I18nAnnotation | string;
}

/**
 * i18n entry
 */
export interface I18nEntry {
    filePath: string;
    key: TextNode;
    value: TextNode;
    /**
     * Annotations for the translation entry, which can contain additional metadata about the entry
     */
    annotation?: I18nAnnotationNode;
}

/**
 * i18n bundle
 */
export type I18nBundle = Record<string, I18nEntry[]>;

/**
 * Text types for texts that are less than 120 characters long
 * https://openui5.hana.ondemand.com/topic/831039835e7c4da3a8a0b49567573afe
 */
export enum SapShortTextType {
    Accessibility = 'XACT',
    AlternativeText = 'XALT',
    BreadcrumbStep = 'XBCB',
    BulletListItemText = 'XBLI',
    ButtonText = 'XBUT',
    Caption = 'XCAP',
    Cell = 'XCEL',
    Checkbox = 'XCKL',
    ColumnHeader = 'XCOL',
    Tabstrip = 'XCRD',
    DataNavigationText = 'XDAT',
    Label = 'XFLD',
    Frame = 'XFRM',
    Term = 'XGLS',
    GroupTitle = 'XGRP',
    Heading = 'XHED',
    LegendText = 'XLGD',
    HyperlinkText = 'XLNK',
    LogEntry = 'XLOG',
    ListBoxItem = 'XLST',
    MenuHeader = 'XMEN',
    MenuItem = 'XMIT',
    MessageText = 'XMSG',
    RadioButton = 'XRBL',
    RoadmapStep = 'XRMP',
    TableRowHeading = 'XROW',
    SelectionText = 'XSEL',
    TabStripText = 'XTBS',
    TableTitle = 'XTIT',
    TreeNodeText = 'XTND',
    QuickInfoText = 'XTOL',
    GeneralText = 'XTXT'
}

/**
 * Text types for texts that are more than 120 characters long
 * https://openui5.hana.ondemand.com/topic/831039835e7c4da3a8a0b49567573afe
 */
export enum SapLongTextType {
    Accessibility = 'YACT',
    BulletListItemText = 'YBLI',
    Definition = 'YDEF',
    Description = 'YDES',
    Explanation = 'YEXP',
    FaqAnswer = 'YFAA',
    Faq = 'YFAQ',
    GlossaryDefinition = 'YGLS',
    Information = 'YINF',
    Instruction = 'YINS',
    LogEntry = 'YLOG',
    ErrorMessage = 'YMSE',
    MessageText = 'YMSG',
    InformationMessageLong = 'YMSI',
    WarningMessage = 'YMSW',
    TechnicalText = 'YTEC',
    Ticker = 'YTIC',
    GeneralTextLong = 'YTXT'
}

export const NOT_RELEVANT_FOR_TRANSLATION = 'NOTR';

export type SapTextType = SapShortTextType | SapLongTextType | typeof NOT_RELEVANT_FOR_TRANSLATION;

/**
 * CDS environment.
 */
export interface CdsEnvironment {
    i18n?: CdsI18nEnv;
}

/**
 * CDS i18n configuration.
 */
export interface CdsI18nConfiguration {
    folders: string[];
    baseFileName: string;
    defaultLanguage: string;
    fallbackLanguage: string;
}

/**
 * CDS i18n environment.
 */
export interface CdsI18nEnv {
    folders?: string[];
    file?: string;
    default_language?: string;
    fallback_bundle?: string;
}
