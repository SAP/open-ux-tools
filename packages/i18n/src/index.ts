export { getCapI18nBundle, getPropertiesI18nBundle } from './read';
export { createCapI18nEntries, createPropertiesI18nEntries } from './write';

export {
    getI18nFolderNames,
    getI18nMaxLength,
    getI18nTextType,
    extractI18nKey,
    extractDoubleCurlyBracketsKey,
    getI18nUniqueKey,
    convertToCamelCase,
    convertToPascalCase,
    printPropertiesI18nEntry,
    printPropertiesI18nAnnotation
} from './utils';

export {
    CdsEnvironment,
    CdsI18nConfiguration,
    CdsI18nEnv,
    I18nAnnotation,
    I18nAnnotationNode,
    I18nBundle,
    I18nEntry,
    NewI18nEntry,
    SapLongTextType,
    SapShortTextType,
    SapTextType,
    ValueNode
} from './types';

export { csvToI18nBundle, jsonToI18nBundle, propertiesToI18nEntry } from './transformer';
