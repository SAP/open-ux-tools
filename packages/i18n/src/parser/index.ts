import { FileFormat } from './types';
import type { ParseResult } from './types';
import { parseProperties } from './properties/parser';
import { parseCsv } from './csv/parser';

/**
 * Parse text.
 *
 * @param text text
 * @param format extension format
 * @returns parse result
 */
export function parse(text: string, format: FileFormat): ParseResult {
    if (format === FileFormat.properties) {
        return parseProperties(text);
    }
    return parseCsv(text);
}
