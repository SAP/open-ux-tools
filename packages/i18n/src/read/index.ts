export { getCapI18nBundle } from './cap';
export { getPropertiesI18nBundle } from './properties';
