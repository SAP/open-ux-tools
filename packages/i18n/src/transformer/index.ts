export { csvToI18nBundle } from './csv';
export { jsonToI18nBundle } from './json';
export { propertiesToI18nEntry } from './properties';
