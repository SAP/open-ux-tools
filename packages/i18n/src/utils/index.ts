export { getI18nConfiguration, getI18nFolderNames } from './config';
export { getCapI18nFiles, resolveCapI18nFolderForFile, getCapI18nFolder } from './resolve';
export { csvPath, jsonPath, capPropertiesPath, doesExist } from './path';
export { printPropertiesI18nAnnotation, printPropertiesI18nEntry } from './print';
export {
    getI18nMaxLength,
    getI18nTextType,
    applyIndent,
    discoverIndent,
    discoverLineEnding,
    convertToCamelCase,
    convertToPascalCase
} from './text';
export { extractI18nKey, getI18nUniqueKey, extractDoubleCurlyBracketsKey } from './key';
export { readFile, writeFile } from './mem-fs-editor';
