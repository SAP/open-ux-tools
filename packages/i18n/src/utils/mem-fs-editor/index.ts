export { readFile } from './read';
export { writeFile } from './write';
