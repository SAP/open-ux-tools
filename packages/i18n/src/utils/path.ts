import { stat } from 'fs';
import type { CdsEnvironment } from '../types';
import { getI18nConfiguration } from './config';

/**
 * Get json path.
 *
 * @param path file path
 * @returns .json file path
 */
export function jsonPath(path: string): string {
    return `${path}.json`;
}
/**
 * Get properties path.
 *
 * @param path file path
 * @param env cds environment
 * @returns .properties file path
 */
export function capPropertiesPath(path: string, env: CdsEnvironment): string {
    const { fallbackLanguage } = getI18nConfiguration(env);
    const languageSuffix = fallbackLanguage === '' ? '' : `_${fallbackLanguage}`;
    return `${path}${languageSuffix}.properties`;
}

/**
 * Get csv path.
 *
 * @param path file path
 * @returns .csv file path
 */
export function csvPath(path: string): string {
    return `${path}.csv`;
}

/**
 * Check if a folder of a file exists.
 *
 * @param path an absolute path to a folder or a file
 * @returns boolean
 */
export function doesExist(path: string): Promise<boolean> {
    return new Promise((resolve) => {
        stat(path, (err) => {
            if (err) {
                resolve(false);
            } else {
                resolve(true);
            }
        });
    });
}
