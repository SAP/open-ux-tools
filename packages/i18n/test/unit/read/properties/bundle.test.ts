import { promises } from 'fs';
import { join } from 'path';
import { getPropertiesI18nBundle } from '../../../../src';
import { replaceBundleWithUnifiedFileUri } from '../../helper';
import { create as createStorage } from 'mem-fs';
import { create } from 'mem-fs-editor';

describe('bundle', () => {
    const DATA_ROOT = join(__dirname, '..', '..', 'data');
    const PROJECT_ROOT = join(DATA_ROOT, 'project');
    test('getPropertiesI18nBundle', async () => {
        const bundle = await getPropertiesI18nBundle(
            join(PROJECT_ROOT, 'app', 'properties-csv', '_i18n', 'i18n.properties')
        );
        replaceBundleWithUnifiedFileUri(PROJECT_ROOT, bundle);
        expect(bundle).toMatchSnapshot();
    });
    test('getPropertiesI18nBundle - mem-fs-editor', async () => {
        const memFs = create(createStorage());
        const bundle = await getPropertiesI18nBundle(
            join(PROJECT_ROOT, 'app', 'properties-csv', '_i18n', 'i18n.properties'),
            memFs
        );
        replaceBundleWithUnifiedFileUri(PROJECT_ROOT, bundle);
        expect(bundle).toMatchSnapshot();
    });
    test('exception / error case', async () => {
        jest.spyOn(promises, 'readFile').mockImplementation(() => {
            throw new Error('should-throw-error');
        });
        const result = getPropertiesI18nBundle(join(PROJECT_ROOT, 'i18n'));
        return expect(result).rejects.toThrowError('should-throw-error');
    });
});
