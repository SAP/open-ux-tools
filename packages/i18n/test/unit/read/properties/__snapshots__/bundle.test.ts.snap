// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle getPropertiesI18nBundle - mem-fs-editor 1`] = `
Object {
  "Properties": Array [
    Object {
      "annotation": undefined,
      "filePath": "app/properties-csv/_i18n/i18n.properties",
      "key": Object {
        "range": Object {
          "end": Object {
            "character": 10,
            "line": 0,
          },
          "start": Object {
            "character": 0,
            "line": 0,
          },
        },
        "value": "Properties",
      },
      "value": Object {
        "range": Object {
          "end": Object {
            "character": 37,
            "line": 0,
          },
          "start": Object {
            "character": 13,
            "line": 0,
          },
        },
        "value": "Properties wins over CSV",
      },
    },
  ],
}
`;

exports[`bundle getPropertiesI18nBundle 1`] = `
Object {
  "Properties": Array [
    Object {
      "annotation": undefined,
      "filePath": "app/properties-csv/_i18n/i18n.properties",
      "key": Object {
        "range": Object {
          "end": Object {
            "character": 10,
            "line": 0,
          },
          "start": Object {
            "character": 0,
            "line": 0,
          },
        },
        "value": "Properties",
      },
      "value": Object {
        "range": Object {
          "end": Object {
            "character": 37,
            "line": 0,
          },
          "start": Object {
            "character": 13,
            "line": 0,
          },
        },
        "value": "Properties wins over CSV",
      },
    },
  ],
}
`;
